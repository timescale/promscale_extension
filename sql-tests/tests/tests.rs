use insta::assert_snapshot;
use std::{
    env,
    mem::MaybeUninit,
    sync::{Mutex, Once},
};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

// TODO fix how test_resources works in nexted workspaces
#[test_resources("sql-tests/testdata/*.sql")]
fn sql_tests(full_resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let resource = if let Some((_, rest)) = full_resource.split_once('/') {
        rest
    } else {
        full_resource
    };
    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
