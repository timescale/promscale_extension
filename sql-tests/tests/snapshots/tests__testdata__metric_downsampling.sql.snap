---
source: sql-tests/tests/tests.rs
expression: query_result
---
 id | schema_name | resolution | retention | should_refresh 
----+-------------+------------+-----------+----------------
(0 rows)

 exists 
--------
 f
(1 row)

DO
CALL
 id | schema_name | resolution | retention | should_refresh 
----+-------------+------------+-----------+----------------
  1 | ds_5m       | 00:05:00   | 30 days   | t
(1 row)

 exists 
--------
 t
(1 row)

 id | downsample_id | metric_id | refresh_pending 
----+---------------+-----------+-----------------
(0 rows)

 id | metric_name | table_name | creation_completed | default_chunk_interval | retention_period | default_compression | delay_compression_until | table_schema | series_table | is_view | view_refresh_interval | downsample_id 
----+-------------+------------+--------------------+------------------------+------------------+---------------------+-------------------------+--------------+--------------+---------+-----------------------+---------------
  1 | test        | test       | f                  | t                      |                  | t                   |                         | prom_data    | test         | f       |                       |              
(1 row)

CALL
 id | metric_name | table_name | creation_completed | default_chunk_interval | retention_period | default_compression | delay_compression_until | table_schema | series_table | is_view | view_refresh_interval | downsample_id 
----+-------------+------------+--------------------+------------------------+------------------+---------------------+-------------------------+--------------+--------------+---------+-----------------------+---------------
  1 | test        | test       | f                  | t                      |                  | t                   |                         | prom_data    | test         | f       |                       |              
  2 | test        | test       | t                  | t                      |                  | t                   |                         | ds_5m        | test         | t       | 00:05:00              |             1
(2 rows)

 id | downsample_id | metric_id | refresh_pending 
----+---------------+-----------+-----------------
  1 |             1 |         1 | f
(1 row)

 update_downsampling_state 
---------------------------
 
(1 row)

 id | schema_name | resolution | retention | should_refresh 
----+-------------+------------+-----------+----------------
  1 | ds_5m       | 00:05:00   | 30 days   | f
(1 row)

 update_downsampling_state 
---------------------------
 
(1 row)

 id | schema_name | resolution | retention | should_refresh 
----+-------------+------------+-----------+----------------
  1 | ds_5m       | 00:05:00   | 30 days   | t
(1 row)

psql:testdata/metric_downsampling.sql:41: NOTICE:  drop cascades to view ds_5m.test
CALL
 id | schema_name | resolution | retention | should_refresh 
----+-------------+------------+-----------+----------------
(0 rows)

 id | downsample_id | metric_id | refresh_pending 
----+---------------+-----------+-----------------
(0 rows)

 id | metric_name | table_name | creation_completed | default_chunk_interval | retention_period | default_compression | delay_compression_until | table_schema | series_table | is_view | view_refresh_interval | downsample_id 
----+-------------+------------+--------------------+------------------------+------------------+---------------------+-------------------------+--------------+--------------+---------+-----------------------+---------------
  1 | test        | test       | f                  | t                      |                  | t                   |                         | prom_data    | test         | f       |                       |              
(1 row)

 exists 
--------
 f
(1 row)


